### DISCUSSION TITLE
Unable to make delta use pager always

### DISCUSSION THREAD

**Discussion by osa1** [dandavison/delta#604] [Q&A] [answered]
Thank you for delta, it's great. I've been trying to make it use pager always when invoked via git. I currently have `--paging=always` everwhere in my .gitconfig where I use delta, but when the diff is small delta still doesn't use paging. Relevant parts of my gitconfig:

```
[core]
    excludesfile = /home/omer/.gitignore_global
    pager = delta --paging=always
[interactive]
    diffFilter = delta --color-only --paging=always
    side-by-side = true
    features = unobstrusive-line-numbers decorations
[delta]
    line-numbers = true
```

Anyone know what the problem could be? Where do I start debugging this? Thanks!

**Comment by dandavison** [comment, marked as answer]
Hey @osa1, thanks!

I think that you can get what you want by doing something like `export DELTA_PAGER='less -R'`.

I am not certain that everything is absolutely perfect regarding the precedence of the various arguments and env vars involved. For example, I personally thought that the following would page, but it did not:

```
echo hello | LESS= DELTA_PAGER= PAGER= delta --paging=always
```

So, there might be some improvements to make to delta! But I think if you are prepared to set that env var then you can get the behaviour you want now.

**Reply by osa1** [reply]
Thanks for the quick response @dandavison . Unfortunately even after adding `export DELTA_PAGER='less -R'` to my .zshrc and creating a new shell (and checking with `echo $DELTA_PAGER`) `git diff` on small diffs still doesn't use paging. I could record a gif if that would be helpful (though I'm not sure how.. I'm on Linux).

**Reply by dandavison** [reply]
Hm, does this use the pager for you? (It does for me.)

```
echo hello | LESS= DELTA_PAGER='less -S' PAGER= delta  --paging=always
```

**Reply by osa1** [reply]
It does, yes.

**Reply by osa1** [reply]
Hm, I thought I installed delta from the git repo, but `--version` says 0.4.2, which is too old. I will update it from the git repo now.

**Reply by osa1** [reply]
Updated with `cargo install --git https://github.com/dandavison/delta` and confirmed with `delta --version`, but delta is still not using a pager when the diff is small.

**Reply by dandavison** [reply]
I don't remember whether the old version would be a problem with this, but it is possible. So, here's the situation as I see it:

1. Is it still possible that the wrong version of delta is being picked up when it is being invoked by git? One way to check is to temporarily set 
```gitconfig
[core]
    pager = delta --version
```

2. We've proved that delta does use the pager when the number of lines is small via the `echo hello | ...` experiment. To convince yourself even more, you could do

```
git diff | LESS= DELTA_PAGER='less -S' PAGER= delta  --paging=always
```

and we expect this to use the pager, even if the diff is small, based on what you observed above with `echo hello | ...`. So, if that is the case, then it must be that somehow delta is getting different values of those environment variables when delta is being invoked by git. You can use the `core.pager` trick again to inspect the environment variables that delta sees when it is invoked by git by doing this:

```gitconfig
[core]
    pager = env
```

**Reply by osa1** [reply]
I tried (1) and it prints 0.7.1. I also tried (2), the env is as expected, and I see a `DELTA_PAGER=less -R` line in the output. Does delta use any other env variables? Perhaps something is overriding the flags/DELTA_PAGER.

**Reply by dandavison** [reply]
Ah, yes, it does also respect `BAT_PAGER`.  (I think I'd like to deprecate this). Do you have that set by any chance?

**Reply by osa1** [reply]
Nope. Here's the full `env` output, with the value removed:

```
USER=...
LANGUAGE=...
LC_TIME=...
XDG_SEAT=...
XDG_SESSION_TYPE=...
SSH_AGENT_PID=...
SHLVL=...
HOME=...
DESKTOP_SESSION=...
XDG_SEAT_PATH=...
LC_MONETARY=...
NIX_PROFILES=...
NIX_PATH=...
DBUS_SESSION_BUS_ADDRESS=...
GLADE_MODULE_PATH=...
MANDATORY_PATH=...
QT_QPA_PLATFORMTHEME=...
LOGNAME=...
DEFAULTS_PATH=...
GTK_OVERLAY_SCROLLING=...
XDG_SESSION_ID=...
CLUTTER_BACKEND=...
PATH=...
GDM_LANG=...
GLADE_PIXMAP_PATH=...
SESSION_MANAGER=...
S_COLORS=...
LC_ADDRESS=...
XDG_RUNTIME_DIR=...
XDG_SESSION_PATH=...
XDG_MENU_PREFIX=...
DISPLAY=...
LANG=...
LC_TELEPHONE=...
XDG_CURRENT_DESKTOP=...
DESKTOP_STARTUP_ID=...
XDG_SESSION_DESKTOP=...
XAUTHORITY=...
XDG_GREETER_DATA_DIR=...
SSH_AUTH_SOCK=...
GLADE_CATALOG_PATH=...
LC_NAME=...
SHELL=...
QT_ACCESSIBILITY=...
GDMSESSION=...
LC_MEASUREMENT=...
NIX_SSL_CERT_FILE=...
GPG_AGENT_INFO=...
LC_IDENTIFICATION=...
XDG_VTNR=...
PWD=...
XDG_DATA_DIRS=...
XDG_CONFIG_DIRS=...
LC_NUMERIC=...
LC_PAPER=...
I3SOCK=...
TERM=...
WINDOWID=...
OLDPWD=...
GHC_BIN=...
EDITOR=...
LD_LIBRARY_PATH=...
LIBRARY_PATH=...
CPLUS_INCLUDE_PATH=...
C_INCLUDE_PATH=...
PKG_CONFIG_PATH=...
PYTHONPATH=...
GOROOT=...
GOPATH=...
TMPDIR=...
FZF_DEFAULT_COMMAND=...
FZF_CTRL_T_COMMAND=...
ORIG_PATH=...
OPAM_SWITCH_PREFIX=...
CAML_LD_LIBRARY_PATH=...
OCAML_TOPLEVEL_PATH=...
MANPATH=...
XDG_CONFIG_HOME=...
RUST_BACKTRACE=...
CARGO_INCREMENTAL=...
DELTA_PAGER=...
_=...
```

**Reply by dandavison** [reply]
No, the relevant env vars are just `LESS`, `PAGER`, `DELTA_PAGER`, and `BAT_PAGER`.

So on my system, when I try to make it page for small diffs, it doesn't immediately work on setting `paging=always`, but it does work when I do this:

```
DELTA_PAGER='less -R' LESS= git -c core.pager='delta --paging=always' diff
```

Can you confirm that that doesn't work for you?

**Reply by osa1** [reply]
That command works! Because I have `DELTA_PAGER` in `env`, and `core.pager` stuff in my .gitconfig, just this much enough to fix this for me: `LESS= git diff`.

Is this expected? Or a bug in delta?

**Reply by dandavison** [reply]
OK, great! So what is the value of `LESS` you have in the environment, before doing `LESS=`?

**Reply by osa1** [reply]
I don't have `LESS` in my env. My full env is shown above.

**Reply by dandavison** [reply]
Right, sorry, yes that does seem undesirable then that setting it to empty string, as opposed to unset, has this rather mysterious consequence! I do think that things could be a bit better here, but it is a bit of a headache doing what users will want under all the different permutations of the env vars and less settings and delta options.

**Reply by osa1** [reply]
Fair enough. Adding `export LESS=` to my .zshrc is good enough for me. Thanks again!

**Comment by simonLeary42** [comment]
alternatively, you can set this variable just for the git pager using `env`:

```
[core]
    pager = env LESS='' delta --paging=always
```
